### PR TITLE
[Capture] add `jit` and `vmap` support to `eval_jaxpr`

### PR DESCRIPTION
**Pull request by albi3ro**
Note that this change isn't necessarily high priority, but I thought I'd just open the PR since I already had the code.

**Context:**

The lightning devices currently throw an error with `jax.jit` and `jax.vmap` and program capture.  While there might be some discussion about where the `jax.pure_callback` should like long term, for now I'm just adding the `jax.pure_callback` call inside the relevant devices.   Not many devices will be implementing `Device.eval_jaxpr` at the moment anyway.

**Description of the Change:**

Adds a `jax.pure_callback` layer in `Device.eval_jaxpr`.

**Benefits:**

`jax.jit` and `jax.vmap` can now be used with capture and lightning devices.

**Possible Drawbacks:**

I don't know if we want to have the `pure_callback` layer live on the device long term.

Also, it turns all errors into `jaxlib.xla_extension.XlaRuntimeError`, which a bit scarier to debug.

**Related Shortcut Stories:**
[sc-83589]

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -40,9 +40,14 @@
...
   [(#1046)](https://github.com/PennyLaneAI/pennylane-lightning/pull/1046)
 
+
+* Capture execution via `dev.eval_jaxpr` can now be used with `jax.jit` and `jax.vmap`.
+  [(#1055)](https://github.com/PennyLaneAI/pennylane-lightning/pull/1055)
+
 * `SX` and `C(SX)` gates are natively supported for all lightning devices.
   [(#731)](https://github.com/PennyLaneAI/pennylane-lightning/pull/731)
 
+
```
Comment (mudit2812) [inline .github/CHANGELOG.md:50]:
```suggestion

* Capture execution via `dev.eval_jaxpr` can now be used with `jax.jit` and `jax.vmap`.
  [(#1055)](https://github.com/PennyLaneAI/pennylane-lightning/pull/1055)

* `SX` and `C(SX)` gates are natively supported for all lightning devices.
  [(#731)](https://github.com/PennyLaneAI/pennylane-lightning/pull/731)

```

Comment (albi3ro) [reply .github/CHANGELOG.md:50]: Getting rid of spaces? I'm having trouble spotting the difference.

File: tests/test_eval_jaxpr.py [hunk 2]
```diff
@@ -446,3 +458,16 @@ def loop(i):
         res = dev.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x, n)
         expected = jax.numpy.cos(0 + 0.1 + 2 * x)
         assert qml.math.allclose(res, expected)
+
+
+def test_vmap_integration():
```
Comment (mudit2812) [inline tests/test_eval_jaxpr.py:463]: Can you also add a test for vmap + jit?

**Review by mudit2812 — COMMENTED** [review]
Ready to approve except for the testing comment.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (JerryChen97) [inline .github/CHANGELOG.md:50]:
```suggestion
```

Comment (albi3ro) [reply .github/CHANGELOG.md]: fixed in local changes.

File: tests/test_eval_jaxpr.py (hunk 2 above)
Comment (albi3ro) [reply tests/test_eval_jaxpr.py]: Added.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (albi3ro) [reply .github/CHANGELOG.md]: Spaces removed.

**Review by JerryChen97 — COMMENTED** [review]
Looks good to me, though probably we also need to hear from Lightning team to see if any concerns regarding the dependecy introduced @josephleekl @maliasadi

File: pennylane_lightning/core/lightning_base.py [hunk 3]
```diff
@@ -490,16 +491,35 @@ def loop(i, y):
 
         """
         # has jax dependency, so can't import up top
-        from .lightning_interpreter import LightningInterpreter
+        import jax
+        from pennylane.capture.primitives import AbstractMeasurement
 
-        if self.wires is None:
-            raise NotImplementedError("Wires must be specified for integration with plxpr capture.")
+        from .lightning_interpreter import LightningInterpreter
```
Comment (JerryChen97) [inline pennylane_lightning/core/lightning_base.py:497]: @josephleekl here is the place where we would like to introduce more dependencies to support `jit` and `vmap`

Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py:497]: What dependency are we adding?

Comment (JerryChen97) [reply pennylane_lightning/core/lightning_base.py:497]: @albi3ro from the context it reads like this file did not expect to be dependent on jax and `lightning_interpreter` seems the place to have it. Not sure if it's totally fine

Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py:497]: I'll reword the comment on 493.

File: tests/test_eval_jaxpr.py [hunk 4]
```diff
@@ -446,3 +458,18 @@ def loop(i):
...
+    """Test that the lightning devices can execute circuits with vmap applied."""
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def circuit(x):
+        qml.RX(x, 0)
+        return qml.expval(qml.Z(0))
+
+    x = jax.numpy.array([1.0, 2.0, 3.0])
+    f = jax.jit(jax.vmap(circuit)) if use_jit else jax.vmap
+    results = jax.vmap(circuit)(x)
```
Comment (mudit2812) [inline tests/test_eval_jaxpr.py:474]:
```suggestion
    f = jax.jit(jax.vmap(circuit)) if use_jit else jax.vmap(circuit)
    results = f(x)
```

**Review by mudit2812 — APPROVED** [review]

File: pennylane_lightning/core/lightning_base.py [hunk 5]
```diff
@@ -490,16 +491,35 @@ def loop(i, y):
 
         """
         # has jax dependency, so can't import up top
```
Comment (albi3ro) [inline pennylane_lightning/core/lightning_base.py:493]:
```suggestion
        # jax is still an optional dependency for pennylane, but mandatory for program capture
        # jax imports cannot be placed in the standard import path
```

File: pennylane_lightning/core/lightning_base.py (hunk 3 above)
Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py]: Mind elaborating? I'm not sure I follow.

**Review by JerryChen97 — APPROVED** [review]
LGTM!

File: pennylane_lightning/core/lightning_base.py (hunk 3 above)
Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py]: Thanks. Add some extra info there too.

File: .github/CHANGELOG.md [hunk 6]
```diff
@@ -44,6 +44,9 @@
 * Reduce the number of shots in the PennyLane Python tests on CIs, from 20k to 10k.
   [(#1046)](https://github.com/PennyLaneAI/pennylane-lightning/pull/1046)
 
+* Capture execution via `dev.eval_jaxpr` can now be used with `jax.jit` and `jax.vmap`.
```
Comment (maliasadi) [inline .github/CHANGELOG.md:47]: Could you add this to the top of the list (line 9)?

File: tests/test_eval_jaxpr.py [hunk 7]
```diff
@@ -80,21 +83,24 @@ def f(x):
...
 
 
 def test_capture_remains_enabled_if_measurement_error():
     """Test that capture remains enabled if there is a measurement error."""
 
-    dev = qml.device(device_name, wires=1)
+    dev = qml.device(device_name, wires=1, shots=1)
 
     def g():
-        return qml.sample(wires=0)  # sampling with analytic execution.
```
Comment (maliasadi) [inline tests/test_eval_jaxpr.py:93]: Is `qml.sample` also supported with `jax.jit`?

File: tests/test_eval_jaxpr.py [hunk 8]
```diff
@@ -461,3 +473,18 @@ def loop(i):
...
+def test_vmap_integration(use_jit):
+    """Test that the lightning devices can execute circuits with vmap applied."""
+
+    @qml.qnode(qml.device(device_name, wires=1))
+    def circuit(x):
+        qml.RX(x, 0)
+        return qml.expval(qml.Z(0))
+
+    x = jax.numpy.array([1.0, 2.0, 3.0])
+    f = jax.jit(jax.vmap(circuit)) if use_jit else jax.vmap(circuit)
```
Comment (maliasadi) [inline tests/test_eval_jaxpr.py:488]: Shall we also test this with non-zero `in_axes` and `out_axes`?

File: pennylane_lightning/core/lightning_base.py [hunk 9]
```diff
@@ -499,9 +500,25 @@ def loop(i, y):
             [1.0, 0.8775825618903728, 0.5403023058681395]
 
         """
-        # has jax dependency, so can't import up top
+        # jax is still an optional dependency for pennylane, but mandatory for program capture
+        # jax imports cannot be placed in the standard import path
+        import jax
```
Comment (maliasadi) [inline pennylane_lightning/core/lightning_base.py:505]: Is this support required a specific version of jax? Do we need to update [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/7d2401500d7f738dcb7d4064ce163c998cee93b8/requirements-dev.txt#L18)?

File: pennylane_lightning/core/lightning_base.py [hunk 10]
```diff
@@ -499,9 +500,25 @@ def loop(i, y):
...
+        from pennylane.capture.primitives import AbstractMeasurement
+
         from .lightning_interpreter import LightningInterpreter
 
+        if jax.config.jax_enable_x64:  # pylint: disable=no-member
+            dtype_map = {
```
Comment (maliasadi) [inline pennylane_lightning/core/lightning_base.py:511]: Could you add a unit test for this branch?

File: pennylane_lightning/core/lightning_base.py (hunk 9 above)
Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py:505]: 0.4.28 is the version currently pinned across the ecosystem.  None of this code should be too opinionated about the version 🤞 Especially compared to some of the stuff we are doing in pennylane and catalyst.

File: tests/test_eval_jaxpr.py (hunk 7 above)
Comment (albi3ro) [reply tests/test_eval_jaxpr.py:93]: Reverting to the no shots + sample case.  Probably changed it when debugging the change in error type.

The point of this test it to trigger some kind of error, and state with finite shots or analytic sampling are equivalent ways to generate an error.

File: pennylane_lightning/core/lightning_base.py (hunk 10 above)
Comment (albi3ro) [reply pennylane_lightning/core/lightning_base.py:511]: tested now.

File: tests/test_eval_jaxpr.py (hunk 8 above)
Comment (albi3ro) [reply tests/test_eval_jaxpr.py:512]: Added test for in_axes and out_axes.

File: .github/CHANGELOG.md (hunk 6 above)
Comment (albi3ro) [reply .github/CHANGELOG.md]: moved.

File: tests/test_eval_jaxpr.py (hunk 7 above)
Comment (albi3ro) [reply tests/test_eval_jaxpr.py:93]: OK now I remember why I switched. With sample and no shots, we get an error when we try and calculate the shapes. 

Here I want to test what happens when an error is thrown mid-execution. So I had to alter what was causing an error.

**Review by AmintorDusko — COMMENTED** [review]
Are these codecov warnings all false positives?

**Comment by albi3ro** [comment]
> Are these codecov warnings all false positives?

I also got them on #1067 as well. These tests are actually running, since I got failures in the CI I had to fix.  So I don't really know what's causing code cov to be so confused.

**Review by maliasadi — APPROVED** [review]
Thanks @albi3ro! Happy to approve
